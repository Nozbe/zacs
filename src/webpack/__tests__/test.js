const webpack = require('webpack')
const { createFsFromVolume, Volume } = require('memfs')
const path = require('path')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')

const compile = (fixture, options = {}) => {
  const compiler = webpack({
    context: __dirname,
    entry: `./${fixture}`,
    output: {
      path: path.resolve(__dirname),
      filename: 'bundle.js',
    },
    plugins: [
      new MiniCssExtractPlugin({
        filename: 'static/[name].css',
        chunkFilename: 'static/[name].[hash].css',
      }),
    ],
    module: {
      rules: [
        {
          test: /\.js$/,
          use: [
            // {
            //   loader: 'babel-loader',
            //   options: {},
            // },
            {
              loader: path.resolve(__dirname, '../loader.js'),
              options,
            },
          ],
        },
        {
          test: /\.css$/,
          use: [
            // {
            //   loader: 'style-loader',
            //   options: {
            //     injectType: 'singletonStyleTag',
            //   },
            // },
            {
              loader: MiniCssExtractPlugin.loader,
              options: {
                // hmr: process.env.NODE_ENV !== 'production',
              },
            },
            {
              loader: 'css-loader',
              options: {
                // importLoaders: 1,
                // sourceMap: process.env.NODE_ENV !== 'production',
              },
            },
          ],
        },
      ],
    },
  })

  compiler.outputFileSystem = createFsFromVolume(new Volume())
  compiler.outputFileSystem.join = path.join.bind(path)

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) {
        reject(err)
      }
      if (stats.hasErrors()) {
        reject(new Error(stats.toJson().errors))
      }

      resolve(stats)
    })
  })
}

describe('zacs-loader', () => {
  it(`extracts CSS`, async () => {
    const stats = await compile('examples/basic.js')
    const generatedModules = stats.toJson().modules
    console.log(generatedModules)
    const output = stats.toJson().modules[0].source
    expect(output).toBe('export default "Hello Alice!"')
  })
})
