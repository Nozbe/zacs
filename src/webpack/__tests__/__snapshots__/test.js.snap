// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zacs-loader extracts CSS 1`] = `
"/* eslint-disable */
/*
ZACS-generated CSS stylesheet begins below.
PRO TIP: If you get a ReferenceError on the line below, it means that your Webpack ZACS support is not configured properly.
If you only see this comment and the one below in generated code, this is normal, nothing to worry about!
*/
const styles = require("../../../../.zacs-cache/src/webpack/__tests__/examples/basic.zacs.css")





/* ZACS-generated CSS stylesheet ends above */

zacs.view(styles.someName)
"
`;

exports[`zacs-loader extracts CSS 2`] = `
"/* Generated CSS file (from ZACS Stylesheets) - do not edit! */

.basic-zacs__someName {
  background-color: red;
  height: 100;
}
"
`;

exports[`zacs-loader extracts CSS 3`] = `
"// extracted by mini-css-extract-plugin
export const someName = "basic-zacs__someName";"
`;

exports[`zacs-loader extracts CSS with slightly malformed whitespace 1`] = `
"/* eslint-disable */
/*
ZACS-generated CSS stylesheet begins below.
PRO TIP: If you get a ReferenceError on the line below, it means that your Webpack ZACS support is not configured properly.
If you only see this comment and the one below in generated code, this is normal, nothing to worry about!
*/
const styles = require("../../../../.zacs-cache/src/webpack/__tests__/examples/brokenWhitespace.zacs.css")











/* ZACS-generated CSS stylesheet ends above */

zacs.view(styles.someName)
"
`;

exports[`zacs-loader has configurable cache directory and extension 1`] = `
"// extracted by mini-css-extract-plugin
export const someName = "basic-zacstest__someName";"
`;
